zkevm-circuits/src/evm_circuit/util/math_gadget.rs

```diff
@@ -820,3 +823,396 @@ impl<F: FieldExt, const N_BYTES: usize> MinMaxGadget<F, N_BYTES> {
         })
     }
 }
+
+// This function generates Lagrange polynomial given a cell, index, and domain
+// size. The polynomial will be equal to 1 when `cell == idx`, otherwise 0.
+// The value of the cell needs to be in the range [0, domain_size)
+fn generate_lagrange_base_polynomial<F: FieldExt>(
+    cell: Cell<F>,
+    idx: u64,
+    domain_size: u64,
+) -> Expression<F> {
+    let mut base_poly = 1.expr();
+    let mut accumulated_inverse = 1.expr();
+    for x in 0..domain_size {
+        if x != idx {
+            base_poly = base_poly * (cell.expr() - x.expr());
+            let inverse = if x < idx {
+                F::from_u128((idx - x) as u128).invert().unwrap()
+            } else {
+                -F::from_u128((x - idx) as u128).invert().unwrap()
+            };
+            accumulated_inverse = accumulated_inverse * inverse;
+        }
+    }
+    base_poly * accumulated_inverse
+}
+
+#[derive(Clone, Debug)]
+pub struct ShrWordsGadget<F> {
+    a: util::Word<F>,
+    shift: util::Word<F>,
+    b: util::Word<F>,
+    // slice_front means the higher part of split digit
+    // slice_back means the lower part of the split digit
+    a_slice_front: [Cell<F>; 32],
+    a_slice_back: [Cell<F>; 32],
```

Comment:
```suggestion
    a_slice_hi: [Cell<F>; 32],
    a_slice_lo: [Cell<F>; 32],
```

Using lower / higher bits are better than front / back.